src/main/java/picocli/CommandLine.java

```diff
@@ -8681,9 +8681,18 @@ public boolean required() {
                 if (desc.length == 0) { return desc; }
                 StringBuilder candidates = new StringBuilder();
                 if (completionCandidates() != null) {
```

Comment:
This call to `completionCandidates()` on line 8683 may also be slow...
It may be better to postpone this call until after we check that `isCompletionCandidatesUsed == true`.

Also, it may be good to avoid calling `completionCandidates()` twice, so how about doing something like this:

```java
if (isCompletionCandidatesUsed) {
    Iterable<String> iter = completionCandidates();
    if (iter != null) {
        for (String c : iter) {
            ....
```

Reply:
Thank you @remkop, and we have updated according to the advice.